engines/e_devcrypto.c

```diff
@@ -772,29 +817,44 @@ static int digest_copy(EVP_MD_CTX *to, const EVP_MD_CTX *from)
         (struct digest_ctx *)EVP_MD_CTX_md_data(to);
     struct cphash_op cphash;
 
-    if (digest_from == NULL || digest_from->init_called != 1)
+    if (digest_from == NULL)
         return 1;
-
-    if (!digest_init(to)) {
+    if (digest_from->inp_count > 0) {
+        digest_to->inp_data = OPENSSL_malloc(digest_from->inp_count);
+        if (digest_to->inp_data == NULL) {
+            fprintf(stderr, "devcrypto: malloc failed\n");
+            goto fail;
+        }
+        memcpy(digest_to->inp_data, digest_from->inp_data, digest_from->inp_count);
+    }
+    if (digest_from->init_called != 1)
```

Comment:
This should be moved with digest_from == NULL above

Reply:
If we move it up, it will return before copying the cached data.  `init_called` is referring to the cryptodev session having been initialized, and we're deferring this to the first update call--to accommodate the HMAC fork case; the rest of the digest changes are a performance booster on their own.